docs/debugging-testing/testing/unit_testing.md

```diff
@@ -1,303 +1,363 @@
+
+All the stub libraries and header libraries are defined under `UNITTESTS/stubs/` directory.
+
+Libraries for fakes are under the `UNITTESTS/fakes` directory. These provide mock implementations that are meant to replace the stub version that does nothing. Usually, these will replace the header files as well as the source files and cannot be used together with their stub equivalents.
+
+The upcoming [Example](#Example) section describes more in detail with writing a unit test for a dummy example class (Unit Under Test), creating example header library, stub example interface library and target_link_libraries into existing mbed-stubs, mbed-headers and also in the unit test CMake configuration file.
+
+### Test discovery
+
+Registering unit tests to run happens automatically, and the test runner handles registration. However, test suites do not automatically build. 
+
+For the build system to build a unit test, pass the `ON` value to `BUILD_TESTING` in the CMake or CTest command-line argument
+
+### Writing unit tests
+
+A unit tests suite consists of one or more test cases. The test cases should cover all the functions in a class under test. All the external dependencies are stubbed including the other classes in the same module. Avoid stubbing header files. Finally, analyze code coverage to ensure all code is tested, and no dead code is found.
+
+Please see the [documentation for Google Test](https://github.com/google/googletest/blob/master/docs/primer.md) to learn how to write unit tests using its framework. See the [documentation for Google Mock](https://github.com/google/googletest/tree/master/googlemock) if you want to write and use C++ mock classes instead of stubs.
+
+#### Test suite configuration
+
+Create two files in the test directory for each test suite:
+
+- Unit test source file (`test_ClassName.cpp`).
+- Unit test CMake configuration file (`CMakeLists.txt`).
+
+List all the required files and libraries for the build in the `CMakeLists.txt` file.
+
+#### Example
+
+With the following steps, you can write a unit test. This example creates dummy classes to be tested, creates and configures unit tests for a class and stubs all external dependencies.
+
+1. Create the following dummy classes header in `mbed-os/example/include/example`:
+
+    **MyClass.h**
+
+    ```
+    #ifndef MYCLASS_H_
+    #define MYCLASS_H_
+
+    namespace example {
+
+    class MyClass {
+    public:
+        int myFunction();
+    };
+
+    }
+
+    #endif
+    ```
+
+1. Add a new `mbed-headers-example` interface library: 
+
+   ### Stub CMake input source file
+
+   ```
+   target_include_directories(mbed-headers-example

```

Comment:
Do we not need to include `add_library(mbed-headers-example INTERFACE)` in the example code snippet?